packages/wrangler/src/experimental-assets.ts

```diff
@@ -1,8 +1,299 @@
+import assert from "node:assert";
 import { existsSync } from "node:fs";
-import path from "node:path";
-import { UserError } from "./errors";
+import { readdir, readFile, stat } from "node:fs/promises";
+import * as path from "node:path";
+import chalk from "chalk";
+import { getType } from "mime";
+import PQueue from "p-queue";
+import prettyBytes from "pretty-bytes";
+import { fetchResult } from "./cfetch";
+import { formatTime } from "./deploy/deploy";
+import { FatalError, UserError } from "./errors";
+import { logger, LOGGER_LEVELS } from "./logger";
+import { hashFile } from "./pages/hash";
+import { isJwtExpired } from "./pages/upload";
+import { APIError } from "./parse";
+import { urlSafe } from "./sites";
 import type { Config } from "./config";
 
+export type AssetManifest = { [path: string]: { hash: string; size: number } };
+
+type InitializeAssetsResponse = {
+	// string of file hashes per bucket
+	buckets: string[][];
+	jwt: string;
+};
+
+export type UploadPayloadFile = {
+	base64: boolean;
+	key: string;
+	value: string;
+	metadata: {
+		contentType: string;
+	};
+};
+
+type UploadResponse = {
+	jwt?: string;
+};
+
+// constants same as Pages for now
+const BULK_UPLOAD_CONCURRENCY = 3;
+const MAX_ASSET_COUNT = 20_000;
+const MAX_ASSET_SIZE = 25 * 1024 * 1024;
+const MAX_UPLOAD_ATTEMPTS = 5;
+const MAX_UPLOAD_GATEWAY_ERRORS = 5;
+
+export const syncExperimentalAssets = async (
+	accountId: string | undefined,
+	scriptName: string,
+	assetDirectory: string | undefined,
+	dryRun: boolean | undefined
+): Promise<string | undefined> => {
+	if (assetDirectory === undefined) {
+		return;
+	}
+	if (dryRun) {
+		logger.log("(Note: doing a dry run, not uploading or deleting anything.)");
+		return;
+	}
+	assert(accountId, "Missing accountId");
+
+	// 1. generate asset manifest
+	logger.info("🌀 Building list of assets...");
+	const manifest = await walk(assetDirectory, {});
+
+	// 2. fetch buckets w/ hashes
+	logger.info("🌀 Starting asset upload...");
+	const initializeAssetsResponse = await fetchResult<InitializeAssetsResponse>(
+		`/accounts/${accountId}/workers/scripts/${scriptName}/assets-upload-session`,
+		{
+			headers: { "Content-Type": "application/json" },
+			method: "POST",
+			body: JSON.stringify({ manifest: manifest }),
+		}
+	);
+
+	// if nothing to upload, return
+	if (initializeAssetsResponse.buckets.flat().length === 0) {
+		if (!initializeAssetsResponse.jwt) {
+			throw new FatalError(
+				"Could not find assets information to attach to deployment. Please try again.",
+				1
+			);
+		}
+		logger.info(`✨ Success! (No files to upload)`);
```

Comment:
```suggestion
		logger.info(`No files to upload. Proceeding with deployment`);
```
or smth along the lines of 👆 . IMHO the `✨ Success!` copy is more relevant when files were actually uploaded. This made me wonder what's the sucess in this case, if no files need be uploaded